File: src/year2024/day6/go/part1.go

package main

import (
	"fmt"
	"os"
	"strings"
)


type Position struct {
    x,y int
}

func main() {
    file, _ := os.ReadFile("../prod1.txt")
    cycle := 0
    grid := strings.Split(strings.TrimSpace(string(file)), "\n")
    visited := make(map[Position]struct{})
    DIRECTIONS := []Position{
        {-1,0},
        {0,1},
        {1,0},
        {0,-1},
    }

    nRows := len(grid)
    nCols := len(grid[0])

    guardPosition := getGuardPosition(grid)
    visited[guardPosition]= struct{}{}

    for {
        currDirection := DIRECTIONS[cycle]
        newPosition := Position{guardPosition.x + currDirection.x, guardPosition.y + currDirection.y}

        if  newPosition.x >= 0  && newPosition.x < nRows &&  newPosition.y >= 0 && newPosition.y < nCols && grid[newPosition.x][newPosition.y] != '#' {
            guardPosition = newPosition
            visited[guardPosition] = struct{}{}
        } else {
            cycle = (cycle + 1) % len(DIRECTIONS)
        }

        if (newPosition.x < 0) || (newPosition.x >= nRows) || (newPosition.y < 0) || (newPosition.y >= nCols) {
            break
        }
    }


    fmt.Printf("Ans: %d", len(visited))
}

func getGuardPosition(grid []string) Position {
    nRows := len(grid)
    nCols := len(grid[0])

    for x := 0; x < nRows; x++ {
        for y:= 0; y < nCols; y++ {
            if string(grid[x][y]) == "^" {
                return Position{x,y}
            }
        }
    }

    return Position{-1,-1}
}

File: src/year2024/day6/scala/part1.scala
import scala.util.control.Breaks._
import scala.collection.mutable.HashSet

@main def part1(): Int = {
  var lines = scala.io.Source.fromFile("../prod1.txt").getLines().toArray
  var grid = lines.map(_.split("").toArray)
  var ans: Int = 0

  // NOTE: remember this syntax
  var visited =  HashSet[(Int, Int)]()
  var cycle = 0
  var nRows = grid.length
  var nCols = grid(0).length
  var guardPosition = (0, 0)

  breakable {
    for (i <- 0 until nRows) {
      for (j <- 0 until nCols) {
        if (grid(i)(j) == "^") {
          guardPosition = (i, j)
          break
        }
      }
    }
  }

  var DIRECTIONS = List(
    (-1, 0),
    (0, 1),
    (1, 0),
    (0, -1),
    )
  visited += guardPosition

  breakable {
    for (i <- 0 until nRows) {
      for (j <- 0 until nCols) {
        var (deltaR, deltaC) = DIRECTIONS(cycle)
        var newR = guardPosition(0) + deltaR
        var newC = guardPosition(1) + deltaC

        if (newR >= 0 && newR < nRows && newC >= 0 && newC < nCols && grid(newR)(newC) != "#") {
          guardPosition = (newR, newC)
          visited += guardPosition
        } else {
          cycle = (cycle + 1) % DIRECTIONS.length
        }

        // We've exited the grid
        if (newR < 0 || newR >= nRows || newC < 0 || newC >= nCols) {
          break
        }
      }
    }
  }


  visited.size
}
